clash-vexriscv-sim/tests/Tests/JtagChain.hs

```diff
@@ -80,70 +77,42 @@ test debug = do
         , std_out = CreatePipe
         }
 
-  withCreateProcess vexRiscvProc $ \_ (fromJust -> simStdOut) _ _ -> do
-    logAHandle <- openFile logAPath ReadMode
-    logBHandle <- openFile logBPath ReadMode
-    let
-      logA0 = SP.fromHandle logAHandle
-      logB0 = SP.fromHandle logBHandle
-      simStdOut0 = SP.fromHandle simStdOut
+  withStreamingFiles2 (logAPath, logBPath) $ \(logA, logB) -> do
+    withCreateProcess vexRiscvProc $ \_ (fromJust -> simStdOut) _ _ -> do
+      waitForLine debug simStdOut "JTAG bridge ready at port 7894"
+
+      expectLine debug logA "[CPU] a"
+      expectLine debug logB "[CPU] b"
 
-    _ <- waitForLineInStream debug simStdOut0 "JTAG bridge ready at port 7894"
+      withCreateProcess openOcdProc $ \_ _ (fromJust -> openOcdStdErr) _ -> do
+        waitForLine debug openOcdStdErr "Halting processor"
 
-    logA1 <- expectLineFromStream debug logA0 "[CPU] a"
-    logB1 <- expectLineFromStream debug logB0 "[CPU] b"
+        withCreateProcess gdbProcA $ \_ _ _ gdbProcHandleA -> do
+          withCreateProcess gdbProcB $ \_ _ _ gdbProcHandleB -> do
+            expectLine debug logA "[CPU] a"
+            expectLine debug logB "[CPU] b"
+            expectLine debug logA "[CPU] b"
+            expectLine debug logB "[CPU] a"
 
-    withCreateProcess openOcdProc $ \_ _ (fromJust -> openOcdStdErr) _ -> do
-      let openOcdStream = SP.fromHandle openOcdStdErr
-      _ <- waitForLineInStream debug openOcdStream "Halting processor"
+            gdbAExitCode <- waitForProcess gdbProcHandleA
+            gdbBExitCode <- waitForProcess gdbProcHandleB
+            ExitSuccess @=? gdbAExitCode
+            ExitSuccess @=? gdbBExitCode
 
-      withCreateProcess gdbProcA $ \_ _ _ gdbProcHandleA -> do
-        withCreateProcess gdbProcB $ \_ _ _ gdbProcHandleB -> do
-          _ <- expectLineFromStream debug logA1 "[CPU] a"
-          _ <- expectLineFromStream debug logB1 "[CPU] b"
-          _ <- expectLineFromStream debug logA1 "[CPU] b"
-          _ <- expectLineFromStream debug logB1 "[CPU] a"
+withStreamingFile :: FilePath -> (Handle -> IO a) -> IO a
+withStreamingFile path f =
+  let tailProc = (proc "tail" ["-n", "0", "-f", path]){std_out = CreatePipe}
+   in withCreateProcess tailProc (\_ (fromJust -> h) _ _ -> f h)
 
-          gdbAExitCode <- waitForProcess gdbProcHandleA
-          gdbBExitCode <- waitForProcess gdbProcHandleB
-          ExitSuccess @=? gdbAExitCode
-          ExitSuccess @=? gdbBExitCode
+withStreamingFiles2 :: (FilePath, FilePath) -> ((Handle, Handle) -> IO a) -> IO a
+withStreamingFiles2 (path0, path1) f = do
+  withStreamingFile path0 $ \h0 ->
+    withStreamingFile path1 $ \h1 ->
+      f (h0, h1)
```

Comment:
Implement as recursive function on lists for when we need 3 files?

Reply:
But then we get "pattern match non-exhaustive" warnings :cry:

Reply:
Maybe `Vec` + `vecToTuple`?